app/scripts/migrations/081.js

```diff
@@ -0,0 +1,68 @@
+import { cloneDeep } from 'lodash';
+import {
+  WALLET_SNAP_PERMISSION_KEY,
+  SnapCaveatType,
+} from '@metamask/rpc-methods';
+
+const version = 81;
+
+/**
+ * Prior to this migration, snap <> dapp permissions were wildcards i.e. `wallet_snap_*`.
+ * Now the permission has been changed to `wallet_snap` and the current snap permissions
+ * that are under wildcards will be added as caveats to a parent `wallet_snap` permission.
+ */
+export default {
+  version,
+  async migrate(originalVersionedData) {
+    const versionedData = cloneDeep(originalVersionedData);
+    versionedData.meta.version = version;
+    const state = versionedData.data;
+    const newState = transformState(state);
+    versionedData.data = newState;
+    return versionedData;
+  },
+};
+
+function transformState(state) {
+  const PermissionController = state?.PermissionController || {};
+
+  const { subjects } = PermissionController;
+
+  const snapPrefix = 'wallet_snap_';
+
+  for (const [subjectName, subject] of Object.entries(subjects)) {
```

Comment:
We're assuming here that `subjects` is an object, which _should_ be true but generally we try to avoid making any assumptions in these migrations just in case the state was corrupted somehow. The migrations should return the state unchanged if we find it violates our assumptions in some way, so that we can properly fix it in a follow-up migration if we see that happen in prod.

Migrating this to TypeScript would make this much easier to do in practice. Our new migration TypeScript template uses the `unknown` type to ensure we don't forget to check our assumptions about state shape.